src/main/java/ikura/Bot.java

```diff
@@ -0,0 +1,234 @@
+// Bot.java
+// Copyright (c) 2020, zhiayang, Apache License 2.0.
+
+package ikura;
+
+import java.util.Map;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Scanner;
+import java.util.Optional;
+import java.util.ArrayList;
+import java.util.stream.Stream;
+import java.util.stream.Collectors;
+import java.util.function.BiConsumer;
+
+import java.util.InputMismatchException;
+import java.util.NoSuchElementException;
+
+import ikura.task.Task;
+import ikura.task.Todo;
+import ikura.task.Event;
+import ikura.task.Deadline;
+
+import ikura.util.Pair;
+import ikura.util.StreamUtils;
+import ikura.util.InvalidInputException;
+
+/**
+ * The main class containing the functionality of the application. Does not own the
+ * tasklist nor the frontend.
+ */
+public class Bot {
+
+    private final TaskList tasks;
+    private final Frontend ui;
+
+    /**
+     * Constructs a new Bot using the given Frontend interface and the given TaskList.
+     *
+     * @param ui    the frontend interface to use.
+     * @param tasks the TaskList to use.
+     */
+    public Bot(Frontend ui, TaskList tasks) {
+        this.tasks  = tasks;
+        this.ui     = ui;
+    }
+
+    /**
+     * Parses a line of user input, and executes the command as appropriate. The
+     * return value specifies whether or not the bot should print the exit message.
+     *
+     * @param str the line user input.
+     * @return true if the bot should continue, false if it should exit.
+     */
+    public boolean processCommand(String str) {
+
+        var sc = new Scanner(str);
+        if (!sc.hasNext()) {
+            return true;
+        }
+
+        var cmd = sc.next().strip();
+
+        if (cmd.equals("bye")) {
+            this.ui.println("goodbye");
+            return false;
+        }
+
+        var input = sc.hasNextLine()
+            ? sc.nextLine().strip()
+            : "";
+
+        this.ui.beginLog();
+
+        Optional.ofNullable(Map.<String, BiConsumer<String, String>>of(
+            "done",     this::cmdDone,
```

Comment:
Perhaps this code could be simplified for readability?

Reply:
how would you suggest to simplify it? to me, it's a lot more clear than a series of switch/cases or even if-else blocks.

Reply:
One suggestion I can think of is to make the map an attribute, and abstract out the lambda as a method with a meaningful method name. That way the code would be much shorter and can be understood more easily and more quickly.